src/System.IO.Pipelines/WritableBufferWriter.cs

```diff
@@ -0,0 +1,82 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+
+namespace System.IO.Pipelines
+{
+    public struct WritableBufferWriter
+    {
+        private WritableBuffer _writableBuffer;
+        private Span<byte> _span;
+
+        public WritableBufferWriter(WritableBuffer writableBuffer)
+        {
+            _writableBuffer = writableBuffer;
+            _span = writableBuffer.Buffer.Span;
+        }
+
+        public Span<byte> Span => _span;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Advance(int count)
+        {
+            _span = _span.Slice(count);
+            _writableBuffer.Advance(count);
+        }
+
+        public void Write(byte[] source)
+        {
+            Write(source, 0, source.Length);
+        }
+
+        public void Write(byte[] source, int offset, int length)
+        {
+            if (length <= _span.Length && length > 0)
```

Comment:
Single test it
```
(uint)length <= (uint)_span.Length
```
If length is less than zero then the `uint` version of it will be greater than `(uint)int.Max`